episodes/03-muscle-connect.md

```diff
@@ -0,0 +1,366 @@
+```
+
+::: callout
+The second argument, which specifies the expected type, is optional. If it is given, MUSCLE will check that the user specified a value of the correct type, and if not raise an exception.
+:::
+
+::: challenge
+
+## Exercise 3: 
+In our example several settings have been hard-coded into the model: the total time to simulate `tmax`, the time step to use `dt`, and the model parameter `k`. Change the code such that we request these settings from the MUSCLE manager.
+
+:::::: solution
+
+```python
+from libmuscle import Instance
+
+def reaction(U_init: np.array) -> np.array:
+    """
+    ...
+    """
+    instance = Instance({
+            Operator.F_INIT: ['initial_state'],       # np.array
+            Operator.O_F: ['final_state']})           # np.array
+
+    while instance.reuse_instance():
+        # begin F_INIT
+        t_max = instance.get_setting('t_max', 'float')
+        dt = instance.get_setting('dt', 'float')
+        k = instance.get_setting('k', 'float')
+
+        U = U_init
+
+        t_cur = 0
+        # end F_INIT
+
+        # begin state_update_loop
+        ...
+```
+
+Note that getting settings needs to happen within the reuse loop; doing it before can lead to incorrect results.
+
+::::::
+:::
+
+## Receiving messages
+After getting our settings, we want to receive the initial state for this submodel on the `initial_state` port. Note that we have declared that port above, and declared it to be associated with the F_INIT operator. During F_INIT, messages can only be received, not sent, so that declaration makes `initial_state` a receiving port.
+
+```python
+    msg = instance.receive('initial_state')
+```
+
+The message that we receive can contain several pieces of information. Here, we are interested in the `data` and `timestamp` attributes. We assume the data to be a grid of floats containing our initial state and the time stamp tells us the simulated time at which this state is valid.
+
+```python
+    U = msg.data.array.copy()
+    t_cur = msg.timestamp
+```
+
+::: callout
+The msg.data attribute holds an object of type libmuscle.Grid, which holds a read-only NumPy array and optionally a list of index names. Here, we take the array and make a copy of it for U, so that we can modify U in our upcoming state update. Without calling `.copy()`, U would end up pointing to the same read-only array, and we’d get an error message if we tried to modify it.
+
+The `timestamp` attribute is a double precision float containing the number of simulated (not wall-clock) seconds since the whole simulation started. 
+:::::::::::
+
+## State update loop
+Last thing is a small change in our state update loop. Instead of checking whether the simulation time does not exceed `t_max`, we have to now add up the time when we started the submodel instance, since we are part of a larger simulation.
+
+```python
+    # begin state_update_loop
+        while t_cur + dt < msg.timestamp + t_max:
```

Comment:
@LourensVeen Isn't it more efficient to leave the state update loop as it was: `while t_cur + dt < t_max`
and make sure that the received timestamp is added to t_cur when a global time stamp is sent out? Is there a good reason to do this addition every time in the while loop?

Reply:
Yes, `msg.timestamp` and `t_max` don't change while we're iterating, so we could do that addition once, after receiving `msg` and before starting the loop.

We cannot add the received timestamp to `t_cur` though, as both are absolute times. Every time we enter the state update loop, we need to run from `msg.timestamp` to `msg.timestamp + t_max`. Does that make sense?

Reply:
I changed it in my latest commit to add the timestamp in the initialisation already, the reuse loop then stays as simple as possible and unchanged from the code we start with. See the latest commit.

Reply:
Okay, yes this looks good.

The pedantic hairsplitter in the back of my mind may want to change t_max to t_end at some point, but it would probably be good to keep this and the MUSCLE3 docs in sync, and it can wait. So let's keep it at this, nice job!